### PR TITLE
[sanitizer_symbolizer] Symbolizer Markup for linux.

### PR DESCRIPTION
**Pull request by avillega**
This change adds support for sanitizer symbolizer markup for linux. For more informationa about symbolzier markup please visit https://llvm.org/docs/SymbolizerMarkupFormat.html

### CODE REVIEW

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 1]
```diff
@@ -248,6 +248,44 @@ static void PrintStackAllocations(StackAllocationsRingBuffer *sa,
...
+  // For symbolizer markup it is just necessary to have dump stack
+  // frames for offline symbolization.
+
+  uptr frames = Min((uptr)flags()->stack_history_size, sa->size());
+  if (const ListOfModules *modules =
+          Symbolizer::GetOrInit()->GetRefreshedListOfModules()) {
+    InternalScopedString modules_res;
+    RenderModules(&modules_res, modules,
+                  /*symbolizer_markup=*/ true);
+    Printf("%s", modules_res.data());
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:271]: Would it be helpful to add an example of how this looks as a comment? Makes it easier to grep for as well.

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 2]
```diff
@@ -425,7 +463,11 @@ void PrintAddressDescription(
       auto *sa = (t == GetCurrentThread() && current_stack_allocations)
                      ? current_stack_allocations
                      : t->stack_allocations();
-      PrintStackAllocations(sa, addr_tag, untagged_addr);
+      if (common_flags()->enable_symbolizer_markup) {
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:466]: I'm confused. Why is the `PrintStackAllocationsMarkup` in the else branch?

File: compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_libcdep.cpp [hunk 3]
```diff
@@ -26,10 +28,10 @@ class StackTraceTextPrinter {
                         InternalScopedString *output,
                         InternalScopedString *dedup_token)
       : stack_trace_fmt_(stack_trace_fmt),
-        frame_delimiter_(frame_delimiter),
-        output_(output),
         dedup_token_(dedup_token),
-        symbolize_(RenderNeedsSymbolization(stack_trace_fmt)) {}
+        symbolize_(RenderNeedsSymbolization(stack_trace_fmt, false)),
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_libcdep.cpp:32]: This doesn't seem right, this seems like a hardcoded false

```
bool RenderNeedsSymbolization(const char *format, bool symbolizer_markup) {
  if (symbolizer_markup) {
    // Online symbolization is never needed for symbolizer markup.
    return false;
  }
```

File: compiler-rt/lib/tsan/rtl/tsan_report.cpp [hunk 4]
```diff
@@ -103,12 +103,22 @@ void PrintStack(const ReportStack *ent) {
     Printf("    [failed to restore the stack]\n\n");
     return;
   }
+
+  if (const ListOfModules *modules =
+          Symbolizer::GetOrInit()->GetRefreshedListOfModules()) {
+    InternalScopedString modules_res;
+    RenderModules(&modules_res, modules,
```
Comment (fmayer) [inline compiler-rt/lib/tsan/rtl/tsan_report.cpp:110]: I would prefer if this and the other caller looked like

```
if (common_flags()->enable_symbolizer_markup) 
   RenderModulesMarkup(...);
```

The indirection doesn't add much and only obscures what is happening.

File: compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp [hunk 5]
```diff
@@ -11,20 +11,140 @@
...
+    }
+
+    buffer->append("{{{module:%d:%s:elf:", renderedModules.size(),
+                   module.full_name());
+    for (uptr i = 0; i < module.uuid_size(); i++) {
+      buffer->append("%02x", module.uuid()[i]);
+    }
+    buffer->append("}}}\n");
+
+    const auto ranges = module.ranges();
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:97]: Why do we need this copy? We don't seem to be touching `module` in the loop.

File: compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp [hunk 6]
```diff
@@ -11,20 +11,140 @@
...
+      // module.base_address = dlpi_addr
+      // range.beg = dlpi_addr + p_vaddr
+      // relative address = p_vaddr = range.beg - module.base_address
+      buffer->append(":%p}}}\n", range.beg - module.base_address());
+    }
+
+    renderedModules.push_back({});
+    RenderedModule &curModule = renderedModules.back();
+    curModule.full_name = internal_strdup(module.full_name());
+    internal_memcpy(curModule.uuid, module.uuid(), module.uuid_size());
```
Comment (fmayer) [inline compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:115]: assert that `curModule.uuid` buffer can hold module.uuid_size?

**Comment by fmayer** [comment]
It don't see any positive tests added, you only changed the existing ones to `false`.

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 7]
```diff
@@ -248,6 +248,44 @@ static void PrintStackAllocations(StackAllocationsRingBuffer *sa,
...
+  InternalScopedString frame_desc;
+  Printf("Previously allocated frames:\n");
+  for (uptr i = 0; i < frames; i++) {
+    const uptr *record_addr = &(*sa)[i];
+    uptr record = *record_addr;
+    if (!record)
+      break;
+    uptr pc_mask = (1ULL << 48) - 1;
+    uptr pc = record & pc_mask;
+    frame_desc.append("  record_addr:0x%zx record:0x%zx",
```
Comment (fmayer) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:283]: Do we want to add a new format to the symbolizer markup to support this data?

**Review by vitalybuka — CHANGES_REQUESTED** [review]
We need tests in test/sanitizer_common with the feature enabled

File: compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_printer.cpp [hunk 8]
```diff
@@ -143,7 +143,12 @@ static const char kDefaultFormat[] = "    #%n %p %F %L";
 
 void RenderFrame(InternalScopedString *buffer, const char *format, int frame_no,
                  uptr address, const AddressInfo *info, bool vs_style,
```
Comment (vitalybuka) [inline compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_printer.cpp:145]: vs_style is ignored?

File: compiler-rt/lib/hwasan/hwasan_report.cpp [hunk 9]
```diff
@@ -248,6 +248,44 @@ static void PrintStackAllocations(StackAllocationsRingBuffer *sa,
...
+
+static void PrintStackAllocationsMarkup(StackAllocationsRingBuffer *sa) {
+  // For symbolizer markup it is just necessary to have dump stack
+  // frames for offline symbolization.
+
+  uptr frames = Min((uptr)flags()->stack_history_size, sa->size());
+  if (const ListOfModules *modules =
+          Symbolizer::GetOrInit()->GetRefreshedListOfModules()) {
+    InternalScopedString modules_res;
+    RenderModules(&modules_res, modules,
```
Comment (vitalybuka) [inline compiler-rt/lib/hwasan/hwasan_report.cpp:269]: Looks like RenderModules is unrelated change.
Can you please extract into a separate patch with justification and tests?

File: compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp [hunk 10]
```diff
@@ -107,7 +234,7 @@ void ReportDeadlySignal(const SignalContext &sig, u32 tid,
                         UnwindSignalStackCallbackType unwind,
                         const void *unwind_context) {}
 
-#if SANITIZER_CAN_SLOW_UNWIND
+#  if SANITIZER_CAN_SLOW_UNWIND
```
Comment (vitalybuka) [inline compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:237]: please precommit format fixes in a separate patch

File: compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp [hunk 11]
```diff
@@ -66,38 +186,45 @@ bool Symbolizer::SymbolizeFrame(uptr addr, FrameInfo *info) { return false; }
...
-  CHECK(!RenderNeedsSymbolization(format));
-  buffer->append(kFormatFrame, frame_no, address);
+  RenderFrameMarkup(buffer, format, frame_no, address, info, vs_style,
+                    strip_path_prefix);
 }
 
 Symbolizer *Symbolizer::PlatformInit() {
-  return new (symbolizer_allocator_) Symbolizer({});
+  IntrusiveList<SymbolizerTool> tools;
+  SymbolizerTool *tool = MarkupSymbolizer::get(&symbolizer_allocator_);
```
Comment (vitalybuka) [inline compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:225]: can this use new() as well?

Comment (vitalybuka) [reply compiler-rt/lib/sanitizer_common/sanitizer_symbolizer_markup.cpp:225]: and I think it's rather should be created inside of  Symbolizer();

**Review by vitalybuka — CHANGES_REQUESTED** [review]

File: compiler-rt/lib/hwasan/hwasan_report.cpp (hunk 9 above)
Comment (avillega) [reply compiler-rt/lib/hwasan/hwasan_report.cpp:269]: `RenderModules` is a related change, I added it as part of stacktrace_printer.h to keep a similar pattern to other `Render*` calls. As other reviewers have noted, it is a unnecessary indirection and I will replace in favor of calling RenderModulesMarkup directly.

File: compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_printer.cpp (hunk 8 above)
Comment (avillega) [reply compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_printer.cpp:145]: RenderFrame does use vs_style. Its is not used in symbolizer markup code.

Comment (vitalybuka) [reply compiler-rt/lib/sanitizer_common/sanitizer_stacktrace_printer.cpp:145]: why it then forwarded into RenderFrameMarkup?
if they are mutually exclusive, would it be better replace two bool flags with enum?

**Comment by avillega** [comment]
Will have to close this PR, seems I messed up by doing a rebase (old workflow habit) will create a new one with comments from this one fixed.
